internal/pbkdf2/pbkdf2.go

```diff
@@ -0,0 +1,145 @@
+// seconds. This is safer than the alternative which is that all rounds are
+// benchmarked (eg, using SHA-256 to produce a 64-byte key) for 1 second, and
+// then it's subsequently possible to run a single round in order to produce
+// 32-bytes of output key material in 500ms.
+func Benchmark(targetDuration time.Duration, hashAlg crypto.Hash) (uint, error) {
+	if !hashAlg.Available() {
+		return 0, errors.New("unavailable digest algorithm")
+	}
+
+	// Start with 1000 iterations.
+	iterationsOut := uint(1000)
+	iterations := iterationsOut
+
+	for i := 1; ; i++ {
+		// time the key derivation
+		duration := timeExecution(&Params{Iterations: iterations, HashAlg: hashAlg})
+		if duration > 0 {
+			// calculate the required number of iterations to return, based on the tested
+			// iterations, measured duration and target duration.
+			newIterationsOut := (int64(iterations) * int64(targetDuration)) / int64(duration)
+			if newIterationsOut > math.MaxInt {
+				return 0, errors.New("iteration count result overflow")
+			}
+			iterationsOut = uint(newIterationsOut)
+		}
+
+		// scale up the number of iterations to test next
+		var scale uint
+		switch {
+		case i > 10:
+			return 0, errors.New("insufficient progress")
+		case duration > 500*time.Millisecond:
+			return iterationsOut, nil
+		case duration <= 62*time.Millisecond:
+			scale = 16
+		case duration <= 125*time.Millisecond:
+			scale = 8
+		case duration <= 250*time.Millisecond:
+			scale = 4
+		default:
+			scale = 2
+		}
+		if math.MaxInt/scale < iterations {
+			// It's only possible to hit this on 32-bit platforms. On
+			// 64-bit platforms, we'll always hit the "insufficient progress"
+			// branch first.
+			return 0, errors.New("test iteration count overflow")
+		}
+		iterations *= scale
+		if int64(iterations) > math.MaxInt64/int64(targetDuration) {
+			return 0, errors.New("iteration count result will overflow")
+		}
+	}
+}
+
+// Params are the key derivation parameters for PBKDF2.
+type Params struct {
+	// Iterations are the number of iterations. This can't be
+	// greater than math.MaxInt.
+	Iterations uint
+
+	// HashAlg is the digest algorithm to use. The algorithm
+	// must be available
+	HashAlg crypto.Hash
+}
+
+// Key derives a key of the desired length from the supplied passphrase and salt,
+// using the supplied parameters.
+//
+// This will return an error if the key length or number of iterations are greater
+// than the maximum value of a signed integer, or the supplied digest algorithm is
+// not available.
+func Key(passphrase string, salt []byte, params *Params, keyLen uint) ([]byte, error) {
+	switch {
+	case params == nil:
+		return nil, errors.New("nil params")
+	case params.Iterations > math.MaxInt:
+		return nil, errors.New("too many iterations")
+	case !params.HashAlg.Available():
+		return nil, errors.New("unavailable digest algorithm")
+	case keyLen > math.MaxInt:
+		return nil, errors.New("invalid key length")
```

Comment:
isn't there anything to check between keyLen and HashAlg?

Reply:
There isn't really any need to do that - one could use any hash algorithm to produce a key of any length, eg, if you request a key of 64 bytes with SHA-256 as the digest and 10000 iterations, it runs 10000 iterations twice to produce 2 blocks of 32-bytes and there's nothing in the design of PBKDF2 that disallows this.